### PR TITLE
Musl/api response assertion strictness

### PR DESCRIPTION
**Pull request by musl**
## Summary of Changes

This pull request makes the assertions for API responses more tolerant of newer firmware changes without sacrificing the ability to fail on breaking changes. It does this by changing the assertion from requiring that the response be the exact set of expected keys to requiring that the request at least have all of the expected keys. This means that newer firmware can return new keys without breaking the library. Firmware changes that remove keys or rename keys will still be detectable.

This pull request also updates inoperative pre-commit hook config, adds python 3.8 to tox testing, and cleans up some string formatting things that were blocking pre-commit hooks and tox tests.

## Test Output

[tox.log](https://github.com/scrool/xled/files/5620943/tox.log)

### CODE REVIEW

**Comment by musl** [comment]
My Gen II strings are happily working with the updates I pushed. :)

File: tox.ini [hunk 1]
```diff
@@ -3,6 +3,7 @@ envlist = py27, py35, py36, py37, linters
 
 [travis]
 python =
+    3.8: py38
```
Comment (scrool) [inline tox.ini:6]: This adds support for Python 3.8. Please update `setup.py` as well to ensure this is publicly visible.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -1,13 +1,9 @@
...
-    rev: stable
+    rev: 20.8b1
     hooks:
     - id: black
-      name: black
-      language: system
-      entry: black
-      types: [python]
-      language_version: python3.6
+      language_version: python3
```
Comment (scrool) [inline .pre-commit-config.yaml:6]: Why this change from explicit python version into generic python3?

File: xled/control.py [hunk 3]
```diff
@@ -432,7 +432,7 @@ def update_firmware(self, stage0, stage1):
                 fw_stage_sums[stage] = xled.security.sha1sum(stage1)
             log.debug("Firmware stage %d SHA1SUM: %r", stage, fw_stage_sums[stage])
             if not fw_stage_sums[stage]:
-                msg = "Failed to compute SHA1SUM for firmware stage %d.".format(stage)
+                msg = "Failed to compute SHA1SUM for firmware stage %d." % (stage)
```
Comment (scrool) [inline xled/control.py:435]: Why this change?

File: xled/control.py [hunk 4]
```diff
@@ -464,12 +464,8 @@ def update_firmware(self, stage0, stage1):
                 assert False
 
         if fw_stage_sums != uploaded_stage_sums:
-            log.error(
-                "Firmware SHA1SUMs: %r != uploaded SHA1SUMs",
-                fw_stage_sums,
-                uploaded_stage_sums,
-            )
-            msg = "Firmware SHA1SUMs doesn't match to uploaded SHA1SUMs.".format(stage)
+            log.error("Firmware SHA1SUMs: %r != uploaded SHA1SUMs", fw_stage_sums)
```
Comment (scrool) [inline xled/control.py:467]: Why this change?

File: xled/control.py [hunk 5]
```diff
@@ -199,7 +199,7 @@ def get_mode(self):
         url = urljoin(self.base_url, "led/mode")
         response = self.session.get(url)
         app_response = ApplicationResponse(response)
-        required_keys = [u"code", u"mode", u"shop_mode"]
+        required_keys = [u"code", u"mode"]
```
Comment (scrool) [inline xled/control.py:202]: Maybe make this part of the followup commit?

File: xled/control.py [hunk 6]
```diff
@@ -197,7 +199,8 @@ def get_mode(self):
         url = urljoin(self.base_url, "led/mode")
         response = self.session.get(url)
         app_response = ApplicationResponse(response)
-        assert sorted(app_response.keys()) == [u"code", u"mode"]
+        required_keys = [u"code", u"mode", u"shop_mode"]
```
Comment (scrool) [inline xled/control.py]: Is `shop_mode` required or not? In the previous commit it was removed and here it was put back.

**Comment by scrool** [comment]
Overall I like these pull request. I would like to revisit some changes in some commits. On the top of inline comments please split 2cc6209290ce3ec40508a99213fccfb96c46dc21 into two commits - one that fixes string formatting and second one that fixes use of Black.

Thanks!

File: tox.ini (hunk 1 above)
Comment (musl) [reply tox.ini:6]: 👍

File: .pre-commit-config.yaml (hunk 2 above)
Comment (musl) [reply .pre-commit-config.yaml:6]: I can put it back, it was useful until I realized I'd need to have more than one python around.

File: xled/control.py (hunk 3 above)
Comment (musl) [reply xled/control.py:435]: I couldn't figure out why the previous line was giving the error and the second form wasn't. I'm happy to put it back if the linter passes it.

File: xled/control.py (hunk 4 above)
Comment (musl) [reply xled/control.py:467]: The second variable wasn't being printed.

File: xled/control.py (hunk 5 above)
Comment (musl) [reply xled/control.py:202]: I can split the commit if you like. What do you want to do?

File: xled/control.py (hunk 6 above)
Comment (musl) [reply xled/control.py]: It caused tests to fail, because it wasn't in the cassette for that test. It's not required for library operation so I took it back out instead of adding it to the cassette.

**Comment by musl** [comment]
Thanks, I'll split the commit and push that back up shortly.

File: xled/control.py (hunk 5 above)
Comment (scrool) [reply xled/control.py:202]: In a followup commit you have improved those asserts to check only required part. Make this change part of that commit please.

File: xled/control.py (hunk 3 above)
Comment (scrool) [reply xled/control.py:435]: Ah, maybe because `(stage)` is pretty much useless in this context and `stage` or `(stage,)` should have been used?

File: xled/control.py (hunk 4 above)
Comment (scrool) [reply xled/control.py:467]: Ah right, good catch!

**Comment by scrool** [comment]
Looks great! I would just like to get that python3.8 support in single commit - so increase list of supported python versions at once. Could you reorder those and squash them?

**Comment by musl** [comment]
Sure.

File: setup.py [hunk 7]
```diff
@@ -50,7 +50,9 @@
     include_package_data=True,
     install_requires=requirements,
     tests_require=tests_requirements,
-    extras_require={"tests": tests_requirements,},  # noqa: E231
+    extras_require={
```
Comment (scrool) [inline setup.py:53]: I have run `black .` and interestingly it reverted this change back to one line. Are you sure you don't override some configuration? (I'm not sure If I don't frankly)

Comment (musl) [reply setup.py:53]: I got a little overzealous following your desire to clean up the history, let me check on it real quick.

Comment (musl) [reply setup.py:53]: I'm not sure what black is doing tbh. My current HEAD has it on 3 lines. I'll push it up and all should be well in a sec.

**Comment by musl** [comment]
That order should be a bit better, as well, in case somebody wants to check out a given branch and do testing. It makes it look like I thought about everything in a rational order, too. :D

**Comment by scrool** [comment]
I'll look into that tomorrow, thanks for now!

File: xled/control.py (hunk 6 above)
Comment (musl) [reply xled/control.py]: In any case, I dropped it from the commit it was introduced, re-ordered, and squashed everything so it's like it was never there.

File: xled/control.py (hunk 3 above)
Comment (musl) [reply xled/control.py:435]: I can monkey around with this and try to make it look the same if you want.

**Comment by musl** [comment]
Thanks! 🎉
